### ISSUE TITLE
Logstash not able to shutdown successfully on ES 2.0 with Kafka input

### ISSUE THREAD

**Issue by ppf2** [logstash-plugins/logstash-input-kafka#49]
sudo /etc/init.d/logstash stop 

```
Killing logstash (pid 28592) with SIGTERM 
Waiting logstash (pid 28592) to die... 
Waiting logstash (pid 28592) to die... 
Waiting logstash (pid 28592) to die... 
Waiting logstash (pid 28592) to die... 
Waiting logstash (pid 28592) to die... 
logstash stop failed; still running. 
```

ps -ef | grep -i java 

```
logstash 28592 1 99 23:21 pts/0 01:36:24 /bin/java -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -Djava.awt.headless=true -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -Djava.io.tmpdir=/var/lib/logstash -Djava.rmi.server.hostname=some_host -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=9994 -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Xmx4g -Xss2048k -Djffi.boot.library.path=/opt/logstash/vendor/jruby/lib/jni -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -Djava.awt.headless=true -XX:CMSInitiatingOccupancyFraction=75 -XX:+UseCMSInitiatingOccupancyOnly -Djava.io.tmpdir=/var/lib/logstash -Djava.rmi.server.hostname=some_host -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=9994 -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Xbootclasspath/a:/opt/logstash/vendor/jruby/lib/jruby.jar -classpath : -Djruby.home=/opt/logstash/vendor/jruby -Djruby.lib=/opt/logstash/vendor/jruby/lib -Djruby.script=jruby -Djruby.shell=/bin/sh org.jruby.Main --1.9 /opt/logstash/lib/bootstrap/environment.rb logstash/runner.rb agent -f /etc/logstash/conf.d -l /var/log/logstash/logstash.log 
```

sudo /etc/init.d/logstash status 

```
logstash is running 
```

End up having to sudo kill -9 28592

tail logstash.log 

All there is in the logs are:
```
{:timestamp=>"2015-11-14T00:02:51.597000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:02:51-08:00", {"input_to_filter"=>18, "filter_to_output"=>1, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:02:56.598000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:02:56-08:00", {"input_to_filter"=>18, "filter_to_output"=>1, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:01.598000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:01-08:00", {"input_to_filter"=>19, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:06.599000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:06-08:00", {"input_to_filter"=>20, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:11.599000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:11-08:00", {"input_to_filter"=>20, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:16.601000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:16-08:00", {"input_to_filter"=>19, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:21.601000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:21-08:00", {"input_to_filter"=>20, "filter_to_output"=>2, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:26.602000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:26-08:00", {"input_to_filter"=>20, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:31.603000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:31-08:00", {"input_to_filter"=>19, "filter_to_output"=>1, "outputs"=>[]}], :level=>:warn} 
{:timestamp=>"2015-11-14T00:03:36.604000-0800", :message=>["INFLIGHT_EVENTS_REPORT", "2015-11-14T00:03:36-08:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
```

The above suggests that the Kafka input is no longer sending anything through, but LS is not able to shutdown successfully.

Logstash config:

```
input {
  kafka {
      zk_connect => "hosts"
      consumer_threads => "1"
      group_id => "group"
      topic_id => "topic"
      tags => "fromKafka"
  }
}

filter {
   if "fromKafka" in [tags]
   {
     metrics
     {
        meter => "events"
        add_tag => "kafka_metric"
        flush_interval => "30"
     }
   }

   if "lumberjack_metric" in [tags] or "syslog_metric" in [tags] or "kafka_metric" in [tags]
   {
     mutate
     {
        add_field => { "host" => "%{message}" }
        remove_field => [ "message" ]
     }
   }
}

output {
   if "lumberjack_metric" in [tags] or "syslog_metric" in [tags] or "kafka_metric" in [tags]
   {
      elasticsearch
      {
         hosts => ["hosts"]
         workers => "2"
      }
   }
   else
   {
      elasticsearch {
      hosts => ["hosts"]
      workers => "10"
      user => "user"
      password => "password"
      }
    }
}
```

**Comment by jordansissel** [comment]
@talevy @jsvd any thoughts?

**Comment by joekiller** [comment]
A verbose log would be helpful.

**Comment by talevy** [comment]
I will take a look at this

**Comment by r-tock** [comment]
I am seeing this issue coupled with high CPU usage, I pasted the log here https://github.com/logstash-plugins/logstash-input-kafka/issues/52

**Comment by Noxshun** [comment]
Hi,

I am also encountering the same problem.

Input only from Kafka
Through some very simple filters.
Output directly to a file.

It seems to happen only when the Kafka topic is empty. If there still are some records/elements in the topic, Logstash successfully shuts down cleanly.

As @r-tock is mentioning, this is propably related to the problem with the high CPU usage.

```
{:timestamp=>"2015-12-21T12:07:02.107000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:02+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:07.109000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:07+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:12.110000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:12+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:17.112000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:17+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:22.113000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:22+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:27.115000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:27+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:32.116000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:32+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:37.118000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:37+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:42.120000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:42+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:47.122000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:47+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:52.123000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:52+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
{:timestamp=>"2015-12-21T12:07:57.125000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:07:57+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn}
```

**Comment by jsvd** [comment]
yes @Noxshun can you try updating the plugin to 2.0.3 using `bin/plugin install --version 2.0.3 logstash-input-kafka`? this version includes the fix https://github.com/logstash-plugins/logstash-input-kafka/pull/54

**Comment by Noxshun** [comment]
Hey.

I am currently using version 2.0.3, the CPU issue is no longer a problem. So that fix worked splendidly.

Here is some more verbose log from me sending SIGINT, until it get stuck, and till I SIGTERM it:

```
{:timestamp=>"2015-12-21T12:28:18.282000+0100", :message=>"SIGINT received. Shutting down the pipeline.", :level=>:warn, :file=>"logstash/agent.rb", :line=>"128", :method=>"execute"}
{:timestamp=>"2015-12-21T12:28:18.285000+0100", :message=>"stopping", :plugin=><LogStash::Inputs::Kafka topic_id=>"logstash", zk_connect=>"10.0.131.60:4002", auto_offset_reset=>"smallest", consumer_threads=>4, codec=><LogStash::Codecs::JSON charset=>"UTF-8">, group_id=>"logstash", reset_beginning=>false, queue_size=>20, rebalance_max_retries=>4, rebalance_backoff_ms=>2000, consumer_timeout_ms=>-1, consumer_restart_on_error=>true, consumer_restart_sleep_ms=>0, decorate_events=>false, fetch_message_max_bytes=>1048576, decoder_class=>"kafka.serializer.DefaultDecoder", key_decoder_class=>"kafka.serializer.DefaultDecoder">, :level=>:debug, :file=>"logstash/inputs/base.rb", :line=>"81", :method=>"do_stop"}
log4j, [2015-12-21T12:28:18.288]  INFO: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], ZKConsumerConnector shutting down
log4j, [2015-12-21T12:28:18.303] DEBUG: kafka.utils.KafkaScheduler: Shutting down task scheduler.
log4j, [2015-12-21T12:28:18.305]  INFO: kafka.consumer.ConsumerFetcherManager: [ConsumerFetcherManager-1450697279275] Stopping leader finder thread
log4j, [2015-12-21T12:28:18.307]  INFO: kafka.consumer.ConsumerFetcherManager$LeaderFinderThread: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab-leader-finder-thread], Shutting down
log4j, [2015-12-21T12:28:18.308]  INFO: kafka.consumer.ConsumerFetcherManager$LeaderFinderThread: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab-leader-finder-thread], Stopped
log4j, [2015-12-21T12:28:18.309]  INFO: kafka.consumer.ConsumerFetcherManager$LeaderFinderThread: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab-leader-finder-thread], Shutdown completed
log4j, [2015-12-21T12:28:18.309]  INFO: kafka.consumer.ConsumerFetcherManager: [ConsumerFetcherManager-1450697279275] Stopping all fetchers
log4j, [2015-12-21T12:28:18.309]  INFO: kafka.consumer.ConsumerFetcherThread: [ConsumerFetcherThread-logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab-0-0], Shutting down
log4j, [2015-12-21T12:28:18.312]  INFO: kafka.consumer.SimpleConsumer: Reconnect due to socket error: java.nio.channels.ClosedByInterruptException
log4j, [2015-12-21T12:28:18.312] DEBUG: kafka.consumer.SimpleConsumer: Disconnecting from osl-logship1.mnemonic.no:9092
log4j, [2015-12-21T12:28:18.312] DEBUG: kafka.consumer.SimpleConsumer: Disconnecting from osl-logship1.mnemonic.no:9092
log4j, [2015-12-21T12:28:18.312] DEBUG: kafka.consumer.SimpleConsumer: Disconnecting from osl-logship1.mnemonic.no:9092
log4j, [2015-12-21T12:28:18.313]  INFO: kafka.consumer.ConsumerFetcherThread: [ConsumerFetcherThread-logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab-0-0], Stopped
log4j, [2015-12-21T12:28:18.313]  INFO: kafka.consumer.ConsumerFetcherThread: [ConsumerFetcherThread-logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab-0-0], Shutdown completed
log4j, [2015-12-21T12:28:18.313] DEBUG: kafka.consumer.SimpleConsumer: Disconnecting from osl-logship1.mnemonic.no:9092
log4j, [2015-12-21T12:28:18.314]  INFO: kafka.consumer.ConsumerFetcherManager: [ConsumerFetcherManager-1450697279275] All connections stopped
log4j, [2015-12-21T12:28:18.315] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Clearing up queue
log4j, [2015-12-21T12:28:18.318] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Cleared queue and sent shutdown command
log4j, [2015-12-21T12:28:18.318] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Clearing up queue
log4j, [2015-12-21T12:28:18.318] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Cleared queue and sent shutdown command
log4j, [2015-12-21T12:28:18.318] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Clearing up queue
log4j, [2015-12-21T12:28:18.319] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Cleared queue and sent shutdown command
log4j, [2015-12-21T12:28:18.319] DEBUG: kafka.consumer.ConsumerIterator: Received the shutdown command
log4j, [2015-12-21T12:28:18.319] DEBUG: kafka.consumer.ConsumerIterator: Received the shutdown command
log4j, [2015-12-21T12:28:18.319] DEBUG: kafka.consumer.ConsumerIterator: Received the shutdown command
log4j, [2015-12-21T12:28:18.319] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Clearing up queue
log4j, [2015-12-21T12:28:18.320] DEBUG: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], Cleared queue and sent shutdown command
log4j, [2015-12-21T12:28:18.320] DEBUG: kafka.consumer.ConsumerIterator: Received the shutdown command
log4j, [2015-12-21T12:28:18.320] DEBUG: org.I0Itec.zkclient.ZkClient: Closing ZkClient...
log4j, [2015-12-21T12:28:18.321]  INFO: org.I0Itec.zkclient.ZkEventThread: Terminate ZkClient event thread.
log4j, [2015-12-21T12:28:18.321] DEBUG: org.I0Itec.zkclient.ZkConnection: Closing ZooKeeper connected to 10.0.131.60:4002
log4j, [2015-12-21T12:28:18.322] DEBUG: org.apache.zookeeper.ZooKeeper: Closing session: 0x151c44b0b140001
log4j, [2015-12-21T12:28:18.323] DEBUG: org.apache.zookeeper.ClientCnxn: Closing client for session: 0x151c44b0b140001
log4j, [2015-12-21T12:28:18.327] DEBUG: org.apache.zookeeper.ClientCnxn: Got notification sessionid:0x151c44b0b140001
log4j, [2015-12-21T12:28:18.328] DEBUG: org.apache.zookeeper.ClientCnxn: Got WatchedEvent state:SyncConnected type:NodeChildrenChanged path:/consumers/logstash/ids for sessionid 0x151c44b0b140001
log4j, [2015-12-21T12:28:18.328] DEBUG: org.I0Itec.zkclient.ZkClient: Received event: WatchedEvent state:SyncConnected type:NodeChildrenChanged path:/consumers/logstash/ids
log4j, [2015-12-21T12:28:18.329] DEBUG: org.apache.zookeeper.ClientCnxn: Reading reply sessionid:0x151c44b0b140001, packet:: clientPath:null serverPath:null finished:false header:: 29,-11  replyHeader:: 29,25227,0  request:: null response:: null
log4j, [2015-12-21T12:28:18.329] DEBUG: org.apache.zookeeper.ClientCnxn: Disconnecting client for session: 0x151c44b0b140001
log4j, [2015-12-21T12:28:18.330]  INFO: org.apache.zookeeper.ZooKeeper: Session: 0x151c44b0b140001 closed
log4j, [2015-12-21T12:28:18.330] DEBUG: org.I0Itec.zkclient.ZkClient: Closing ZkClient...done
log4j, [2015-12-21T12:28:18.330] DEBUG: org.I0Itec.zkclient.ZkClient: ignoring event '{NodeChildrenChanged | /consumers/logstash/ids}' since shutdown triggered
log4j, [2015-12-21T12:28:18.330] DEBUG: org.apache.zookeeper.ClientCnxn: An exception was thrown while closing send thread for session 0x151c44b0b140001 : Unable to read additional data from server sessionid 0x151c44b0b140001, likely server has closed socket
log4j, [2015-12-21T12:28:18.330] DEBUG: org.I0Itec.zkclient.ZkClient: Leaving process event
log4j, [2015-12-21T12:28:18.330]  INFO: org.apache.zookeeper.ClientCnxn: EventThread shut down
log4j, [2015-12-21T12:28:18.331]  INFO: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], ZKConsumerConnector shutdown completed in 42 ms
log4j, [2015-12-21T12:28:18.386]  INFO: kafka.consumer.ZookeeperConsumerConnector: [logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab], stopping watcher executor thread for consumer logstash_osl-logstash1.mnemonic.no-1450697279179-a736f0ab
{:timestamp=>"2015-12-21T12:28:21.176000+0100", :message=>"Flushing", :plugin=><LogStash::Filters::Metrics meter=>["events"], add_tag=>["metric"], flush_interval=>60, periodic_flush=>false, ignore_older_than=>0, clear_interval=>-1, rates=>[1, 5, 15], percentiles=>[1, 5, 10, 90, 95, 99, 100]>, :level=>:debug, :file=>"(eval)", :line=>"16", :method=>"initialize"}
{:timestamp=>"2015-12-21T12:28:23.286000+0100", :message=>"Received shutdown signal, but pipeline is still waiting for in-flight events\nto be processed. Sending another ^C will force quit Logstash, but this may cause\ndata loss.", :level=>:warn, :file=>"logstash/agent.rb", :line=>"129", :method=>"execute"}
{:timestamp=>"2015-12-21T12:28:23.289000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:28:23+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn, :file=>"logstash/util/reporter.rb", :line=>"26", :method=>"report"}
{:timestamp=>"2015-12-21T12:28:26.177000+0100", :message=>"Flushing", :plugin=><LogStash::Filters::Metrics meter=>["events"], add_tag=>["metric"], flush_interval=>60, periodic_flush=>false, ignore_older_than=>0, clear_interval=>-1, rates=>[1, 5, 15], percentiles=>[1, 5, 10, 90, 95, 99, 100]>, :level=>:debug, :file=>"(eval)", :line=>"16", :method=>"initialize"}
{:timestamp=>"2015-12-21T12:28:28.291000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:28:28+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn, :file=>"logstash/util/reporter.rb", :line=>"26", :method=>"report"}
{:timestamp=>"2015-12-21T12:28:31.178000+0100", :message=>"Flushing", :plugin=><LogStash::Filters::Metrics meter=>["events"], add_tag=>["metric"], flush_interval=>60, periodic_flush=>false, ignore_older_than=>0, clear_interval=>-1, rates=>[1, 5, 15], percentiles=>[1, 5, 10, 90, 95, 99, 100]>, :level=>:debug, :file=>"(eval)", :line=>"16", :method=>"initialize"}
{:timestamp=>"2015-12-21T12:28:33.293000+0100", :message=>["INFLIGHT_EVENTS_REPORT", "2015-12-21T12:28:33+01:00", {"input_to_filter"=>0, "filter_to_output"=>0, "outputs"=>[]}], :level=>:warn, :file=>"logstash/util/reporter.rb", :line=>"26", :method=>"report"}
{:timestamp=>"2015-12-21T12:28:34.798000+0100", :message=>"SIGINT received. Terminating immediately..", :level=>:fatal, :file=>"logstash/agent.rb", :line=>"125", :method=>"execute"}
```

**Comment by jsvd** [comment]
I'm not very familiar with the kafka plugins, maybe @talevy can lend a hand here?

**Comment by joekiller** [comment]
I think i had the loop as it was prior to the cpu fix due to blocking on
the queue and in ability for the plugin to shut down. We'll have to examine
more.

**Comment by joekiller** [comment]
I said this on #62 but I'll paste it here, with the new jruby-kafka consumer design I think we can fix this. The consumption will move into a different thread and we can move some logic into the thread as @jsvd described to push a message to shutdown the process.

**Comment by suyograo** [comment]
@joekiller Did you recently publish a new jruby-kafka consumer? Would you be willing to do a PR for this?

**Comment by suyograo** [comment]
Fixed in https://github.com/logstash-plugins/logstash-input-kafka/pull/69 by @talevy 

available in v2.0.4
